src/transformers/models/llama/tokenization_llama.py

```diff
@@ -117,9 +121,33 @@ def get_vocab(self):
         vocab.update(self.added_tokens_encoder)
         return vocab
 
+    # Copied from transformers.models.t5.tokenization_t5.T5Tokenizer.tokenize
+    def tokenize(self, text, **kwargs) -> List[str]:
+        # Replace the SPIECE_UNDERLINE with a space to make sure SPIECE_UNDERLINE is only used at
+        # the beginning of the text
+        if not self.legacy:
+            text = SPIECE_UNDERLINE + text.replace(SPIECE_UNDERLINE, " ")
+        return super().tokenize(text, **kwargs)
+
+    # Copied from transformers.models.t5.tokenization_t5.T5Tokenizer._tokenize
     def _tokenize(self, text):
-        """Returns a tokenized string."""
-        return self.sp_model.encode(text, out_type=str)
+        """
+        Returns a tokenized string. Since the sentencpiece internal model always adds a SPIECE_UNDERLINE, at the
+        beginning of the provided text, we need to remove it by hand when the current text is a subsequence. This
+        happens whenever the `self.tokenize` function is called with specials tokens: the input is split on the special
+        tokens, and each subsequence is passed to `_tokenize`. Thus is a subsequence did not start with a `" "` or
+        SPIECE_UNDERLINE, we have to remove the extra `SPIECE_UNDERLINE` prepended.
+        """
+        if not self.legacy:
+            is_first = text.startswith(SPIECE_UNDERLINE)
+            if is_first:
+                text = text[1:]
+
+        tokens = self.sp_model.encode(text, out_type=str)
+
+        if not self.legacy and not is_first and not text.startswith(" ") and tokens[0].startswith(SPIECE_UNDERLINE):
+            tokens = ([tokens[0][1:]] if len(tokens[0]) > 1 else []) + tokens[1:]
```

Comment:
I don't believe you need the check here, I can do a slice with an index not in a list e.g.:
```python
li = [0]
print(li[2:])
```

```suggestion
            tokens = tokens[0][1:] + tokens[1:]
```

Reply:
Actually this does not work 😅 
The reason is that when the token's length is exactly one, we get a string. Even if we wrap it around `[tokens[0][1:]] + tokens[1:]`, then there is a problem when the string is null length, it will be `[[]]`. So reverted this and left what was here! I'm pretty sure I tried both and ended up having to add the complexity!